File: sdk/typescript/src/volumes.ts
import { CommandFailedError } from "./errors.ts";
import { runCli } from "./process.ts";
import { CommandResult } from "./shell.ts";
import type { VolumeInfo } from "./types.ts";

function mapVolume(row: Record<string, unknown>): VolumeInfo {
  return {
    name: String(row.name),
    guest: (row.guest as string | null) ?? null,
    base: (row.base as string | null) ?? null,
    path: String(row.path),
    size: String(row.size),
    createdAt: row.created_at == null ? null : Number(row.created_at),
    tracked: Boolean(row.tracked),
  };
}

/** List persistent volumes. */
export async function listVolumes(): Promise<VolumeInfo[]> {
  const res = await runCli(["volume", "ls", "--json"]);
  if (res.exitCode !== 0) {
    throw new CommandFailedError(
      new CommandResult(res.stdout, res.stderr, res.exitCode, "bsdkrun volume ls"),
    );
  }
  const rows = JSON.parse(res.stdout || "[]") as Record<string, unknown>[];
  return rows.map(mapVolume);
}

/** Remove one or more volumes (and their data). */
export async function removeVolume(
  names: string | string[],
  opts: { force?: boolean } = {},
): Promise<void> {
  const list = Array.isArray(names) ? names : [names];
  const args = ["volume", "rm"];
  if (opts.force) args.push("--force");
  args.push(...list);
  const res = await runCli(args);
  if (res.exitCode !== 0) {
    throw new CommandFailedError(
      new CommandResult(res.stdout, res.stderr, res.exitCode, "bsdkrun volume rm"),
    );
  }
}

/** Namespace grouping volume operations. */
export const volumes = {
  list: listVolumes,
  remove: removeVolume,
};

File: sdk/rust/src/volumes.rs
//! Host-level volume operations.

use serde_json::Value;

use crate::error::Result;
use crate::process::run_checked;
use crate::types::VolumeInfo;

/// List persistent volumes.
pub fn list() -> Result<Vec<VolumeInfo>> {
    let res = run_checked(["volume", "ls", "--json"], "bsdkrun volume ls")?;
    let raw = if res.stdout.trim().is_empty() {
        "[]".to_string()
    } else {
        res.stdout
    };
    let rows: Value = serde_json::from_str(&raw)?;
    Ok(rows
        .as_array()
        .map(|rows| rows.iter().map(VolumeInfo::from_row).collect())
        .unwrap_or_default())
}

/// Remove one or more volumes (and their data). `force` removes them even
/// while referenced.
pub fn remove<S: AsRef<str>>(names: &[S], force: bool) -> Result<()> {
    let mut args = vec!["volume".to_string(), "rm".to_string()];
    if force {
        args.push("--force".to_string());
    }
    args.extend(names.iter().map(|n| n.as_ref().to_string()));
    run_checked(args, "bsdkrun volume rm")?;
    Ok(())
}
